cache middleware using redis

AchmadQiz opened the pull request:
add caching using redis in the 'get all review' endpoint. you need to install redis on your local machine first and set up the redis configuration in the environment file. If not configured, then automatically use the default url : redis://localhost:6379/0.

here is the benchmark comparison of performing 2000 requests with 200 requests concurrently

without using caching
![Screenshot 2024-02-21 001416](https://github.com/project-zusammen/hidden-gem-backend/assets/113181501/a1d12b23-42d8-4cca-894e-7d4abaa276e6)

using caching
![Screenshot 2024-02-21 001358](https://github.com/project-zusammen/hidden-gem-backend/assets/113181501/f0c51de9-269c-4ede-9cb8-d937a9d1f90b)

fatjan reviewed (dismissed):
awesome, thanks for this PR.

fatjan commented on app/settings.py:
```diff
@@ -25,3 +25,14 @@ def get_db_uri(db_user, db_password):
+REDIS_HOST = env.str("REDIS_HOST", default="localhost")
+REDIS_PORT = env.int("REDIS_PORT", default=6379)
+CACHE_REDIS_URL = f"redis://{REDIS_HOST}:{REDIS_PORT}/0"
+CACHE_TYPE = env.str("CACHE_TYPE", default="redis" if CACHE_REDIS_URL else "simple")
```
I think the if else is unnecessary, no?

the `bool(CACHE_REDIS_URL)` will always be `True`, right?

AchmadQiz replied:
thats right, i have to update it

AchmadQiz replied:
what if we don't need to provide default values for the redis host and redis port so if we want to use redis, we have to add it in the environment variables. If it's empty, then it will use the 'simple' cache type. 
what do you think?

AchmadQiz commented:
i have updated the configuration where if the redis host and redis port are not defined in the environment variables, it will default to using the simple cache type. the benchmark results are almost the same. also, i have adjusted the redis url based on the redis credentials, whether local or cloud.
and please add credentials to the environment variables in the pipeline. i don't have access to add github secrets

fatjan commented on app/settings.py:
```diff
@@ -25,3 +25,22 @@ def get_db_uri(db_user, db_password):
+if REDIS_HOST and REDIS_PORT:
+    if REDIS_USERNAME and REDIS_PASSWORD:
+        redis_cache_config["CACHE_TYPE"] = "redis"
+        redis_cache_config["CACHE_REDIS_URL"] = f"redis://:{REDIS_PASSWORD}@{REDIS_HOST}:{REDIS_PORT}/0"
```
Does this mean the `REDIS_USERNAME` is not needed?

AchmadQiz replied:
well, yes
